ddspipe_core/src/cpp/monitoring/producers/TopicsMonitorProducer.cpp

```diff
@@ -0,0 +1,333 @@
+
+    for (auto& consumer : consumers_)
+    {
+        consumer->consume(&data);
+    }
+
+    reset_data_();
+}
+
+void TopicsMonitorProducer::msg_received(
+        const types::DdsTopic& topic,
+        const types::ParticipantId& participant_id)
+{
+    if (!enabled_)
+    {
+        // Don't save the data if the producer is not enabled
+        return;
+    }
+
+    // Take the lock to prevent:
+    //      1. Changing the data while it's being saved.
+    //      2. Simultaneous calls to msg_received.
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    const auto& topic_name = topic.m_topic_name;
+    const auto& type_name = topic.type_name;
+
+    if (!topic_data_.count(topic_name))
+    {
+        topic_data_[topic_name].name(topic_name);
+        topic_data_[topic_name].type_name(type_name);
+    }
+
+    if (!participant_data_.count(topic_name) || !participant_data_[topic_name].count(participant_id))
+    {
+        // First message received for topic & participant
+        // Save the participant_id
+        participant_data_[topic_name][participant_id].participant_id(participant_id);
+    }
+
+    // Increase the count of the received messages
+    participant_data_[topic_name][participant_id].msgs_received(participant_data_[topic_name][participant_id].msgs_received() +
+            1);
+}
+
+void TopicsMonitorProducer::msg_lost(
+        const types::DdsTopic& topic,
+        const types::ParticipantId& participant_id)
+{
+    if (!enabled_)
+    {
+        // Don't save the data if the producer is not enabled
+        return;
+    }
+
+    // Take the lock to prevent:
+    //      1. Changing the data while it's being saved.
+    //      2. Simultaneous calls to msg_lost.
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    const auto& topic_name = topic.m_topic_name;
+    const auto& type_name = topic.type_name;
+
+    if (!participant_data_.count(topic_name) || !participant_data_[topic_name].count(participant_id))
+    {
+        // First message lost for topic & participant
+        // Save the participant_id
+        participant_data_[topic_name][participant_id].participant_id(participant_id);
+    }
+
+    // Increase the count of the lost messages
+    participant_data_[topic_name][participant_id].msgs_lost(participant_data_[topic_name][participant_id].msgs_lost() + 1);
+}
+
+void TopicsMonitorProducer::type_discovered(
+        const std::string& topic_name,
+        const std::string& type_name)
+{
+    if (!enabled_)
+    {
+        // Don't save the data if the producer is not enabled
+        return;
+    }
+
+    // Take the lock to prevent:
+    //      1. Changing the data while it's being saved.
+    //      2. Simultaneous calls to msg_lost.
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    if (!topic_data_.count(topic_name))
+    {
+        topic_data_[topic_name].name(topic_name);
+        topic_data_[topic_name].type_name(type_name);
+    }
+
+    topic_data_[topic_name].type_discovered(true);
```

Comment:
We should think if the topic name is the right value for the key. I mean, the DDS-Pipe (almost surely) supports two topics with same topic name but different type name.

Reply:
I would like to make `DdsTopic` the key of the map. The problem is that in `internal_notify_type_object_` I only have access to the `type_name`. I thought I also had access to the `topic_name` but I don't.

Reply:
Way better now.